src/ErgoDex/Contracts/Proxy/PlutarchSwap.hs

```diff
@@ -0,0 +1,207 @@
+
+  pure $ isTxSignedByRewardKey #|| (validPool #&& validNumInputs #&& validRewardProp #&& fairExFee #&& correctQuote #&& fairPrice)
+
+pFairExFee 
+  :: Term s (
+         PValue 
+    :--> PValue 
+    :--> PAssetClass 
+    :--> PInteger 
+    :--> PAssetClass 
+    :--> PInteger 
+    :--> PInteger 
+    :--> PInteger 
+    :--> PBool
+    )
+pFairExFee = plam $ \rewardValue orderValue base baseAmount quote quoteAmount exFeePerTokenNum exFeePerTokenDen ->
+  let
+    bqAda  = 
+      pif (pIsAda # base) 
+        (ptuple # pdata baseAmount # zeroInteger)
+        (pif (pIsAda # quote) (ptuple # zeroInteger # pdata quoteAmount) (ptuple # zeroInteger # zeroInteger))
+    baseAda  = pfromData $ pfield @"_0" # bqAda
+    quoteAda = pfromData $ pfield @"_1" # bqAda
+    outAda = pGetLovelace # rewardValue
+    inAda  = pGetLovelace # orderValue
+    exFee  = pdiv # (quoteAmount * exFeePerTokenNum) # exFeePerTokenDen
+  in 
+    (inAda - baseAda - exFee) #< (outAda - quoteAda)
+  
+pFairPrice 
+  :: Term s (
+         PInteger 
+    :--> PValue 
+    :--> PAssetClass 
+    :--> PAssetClass 
+    :--> PInteger 
+    :--> PInteger 
+    :--> PBool
+    )
+pFairPrice = plam $ \quoteAmount poolValue base quote feeNum baseAmount ->
+  let
+    relaxedOut    = quoteAmount + 1
+    reservesBase  = assetClassValueOf # poolValue # base
+    reservesQuote = assetClassValueOf # poolValue # quote
+    feeDen        = 1000
+  in
+    reservesQuote * baseAmount * feeNum #<= relaxedOut * (reservesBase * feeDen + baseAmount * feeNum)
+
+pQuoteAmount :: Term s (PValue :--> PValue :--> PAssetClass :--> PInteger :--> PInteger :--> PInteger)
+pQuoteAmount = plam $ \rewardValue orderValue quote exFeePerTokenDen exFeePerTokenNum ->
+  let
+    quoteOut   = assetClassValueOf # rewardValue # quote
+    quoteIn    = assetClassValueOf # orderValue # quote
+    quoteDelta = quoteOut - quoteIn
+  in
+    pif (pIsAda # quote) (pdiv # (quoteDelta * exFeePerTokenDen) # (exFeePerTokenDen - exFeePerTokenNum)) quoteDelta  
+
+pAdaCS :: Term s PCurrencySymbol
+pAdaCS = pcon $ PCurrencySymbol $ phexByteStr ""
+
+pAdaTN :: Term s PTokenName
+pAdaTN = pcon $ PTokenName $ phexByteStr ""
+
+pAdaAC :: Term s PAssetClass
+pAdaAC = phoistAcyclic $
+    pcon $
+      PAssetClass $
+        pdcons
+          # pdata pAdaCS #$ pdcons
+          # pdata pAdaTN #$ pdnil
+
+pIsAda :: Term s (PAssetClass :--> PBool)
+pIsAda = plam $ \ac -> ac #== pAdaAC
+
+pGetLovelace :: Term s (PValue :--> PInteger)
+pGetLovelace = plam $ \value ->
+  assetClassValueOf # value # pAdaAC
+
+zeroInteger :: Term s (PAsData PInteger)
+zeroInteger = pdata 0
+
+pToPubKeyHash :: Term s (PAddress :--> PPubKeyHash)
+pToPubKeyHash = plam $ \address ->
+  let credential = pfromData $ pfield @"credential" # address
+  in pmatch credential $ \case
+        PPubKeyCredential x -> pfromData $ pfield @"_0" # x
+        _ -> perror
+
+pTxSignedBy :: Term s (PBuiltinList (PAsData PPubKeyHash) :--> PPubKeyHash :--> PBool)
```

Comment:
to check if user pubkeyhash exists in signatures you can use pelem

Reply:
done